apps/smime.c

```diff
@@ -607,13 +609,17 @@ int smime_main(int argc, char **argv)
         }
     } else if (operation == SMIME_VERIFY) {
         STACK_OF(X509) *signers;
-        if (PKCS7_verify(p7, other, store, indata, out, flags))
+        int i;
+
+        for (i = 0; i < sk_X509_num(untrusted_certs); i++)
+            PKCS7_add_certificate(p7, sk_X509_value(untrusted_certs, i));
```

Comment:
What is the point of adding these to the `p7` object? It is not clear what's wrong with the original line `610`. In both cases `NOINTERN` will use only the `-certfile` certs, and otherwise both what's in the message and the additional chain certs.

Reply:
The idea was that the certs given with `-certfile` should be made available for chain building during verification.
Yet as you pointed out, doing it this way has an unwanted side effect on the set of certs that are candidates for signer certs.

Yesterday I did not mention that I was already planning to replace this problematic part,
namely by new code in `PKCS7_verify()` that instead combines on-the-fly the certs given with `-certfile` with the certs contained in the message for for use with chain building.
I just didn't have the time to do so, but likely very soon...